camayoc/tests/qcs/cli/test_sources.py

```diff
@@ -0,0 +1,879 @@
+        [(CONNECTION_PASSWORD_INPUT, utils.uuid4())],
+    )
+
+    qpc_source_add = pexpect.spawn(
+        'qpc source add --name {} --cred {} --hosts {} --type network'
+        .format(name, cred_name, hosts)
+    )
+    assert qpc_source_add.expect('Source "{}" was added'.format(name)) == 0
+    assert qpc_source_add.expect(pexpect.EOF) == 0
+    qpc_source_add.close()
+    assert qpc_source_add.exitstatus == 0
+
+    if hosts.endswith('0/24'):
+        hosts = hosts.replace('0/24', '\[0:255\]')
+    source_show(
+        {'name': name},
+        '{{\r\n'
+        '    "credentials": \[\r\n'
+        '        {{\r\n'
+        '            "id": \\d+,\r\n'
+        '            "name": "{}"\r\n'
+        '        }}\r\n'
+        '    \],\r\n'
+        '    "hosts": \[\r\n'
+        '        "{}"\r\n'
+        '    \],\r\n'
+        '    "id": \\d+,\r\n'
+        '    "name": "{}",\r\n'
+        '    "port": 22,\r\n'
+        '    "source_type": "network"\r\n'
+        '}}\r\n'
+        .format(cred_name, hosts, name)
+    )
+
+
+@pytest.mark.parametrize('hosts', ('127.0.0.1', '192.168.0.0/24'))
+def test_add_with_cred_hosts_file(
+        isolated_filesystem, qpc_server_config, hosts):
+    """Add a source with cred and hosts populated on a file.
+
+    :id: 93d10834-9e9a-4713-8786-918d3d87a4b0
+    :description: Add a source entry providing the ``--name``, ``--cred`` and
+        ``--hosts`` options, the value of the ``--hosts`` options should be a
+        file.
+    :steps: Run ``qpc source add --name <name> --cred <cred> --hosts
+        <hosts_file> --type <type>``
+    :expectedresults: A new source entry is created with the data provided as
+        input.
+    """
+    cred_name = utils.uuid4()
+    name = utils.uuid4()
+    cred_add(
+        {
+            'name': cred_name,
+            'username': utils.uuid4(),
+            'password': None,
+        },
+        [(CONNECTION_PASSWORD_INPUT, utils.uuid4())],
+    )
+
+    with open('hosts_file', 'w') as handler:
+        handler.write(hosts + '\n')
+
+    qpc_source_add = pexpect.spawn(
+        'qpc source add --name {} --cred {} --hosts {} --type network'
+        .format(name, cred_name, 'hosts_file')
+    )
+    assert qpc_source_add.expect('Source "{}" was added'.format(name)) == 0
+    assert qpc_source_add.expect(pexpect.EOF) == 0
+    qpc_source_add.close()
+    assert qpc_source_add.exitstatus == 0
+
+    if hosts.endswith('0/24'):
+        hosts = hosts.replace('0/24', '\[0:255\]')
+    source_show(
+        {'name': name},
+        '{{\r\n'
+        '    "credentials": \[\r\n'
+        '        {{\r\n'
+        '            "id": \\d+,\r\n'
+        '            "name": "{}"\r\n'
+        '        }}\r\n'
+        '    \],\r\n'
+        '    "hosts": \[\r\n'
+        '        "{}"\r\n'
+        '    \],\r\n'
+        '    "id": \\d+,\r\n'
+        '    "name": "{}",\r\n'
+        '    "port": 22,\r\n'
+        '    "source_type": "network"\r\n'
+        '}}\r\n'
```

Comment:
I forget where you landed on this last time with a similar issue, but you may consider if this regex can get wrapped up in a constant or constructed by a helper function. Also, you may have mentioned your intention to do this in the chat.

Reply:
This is something that I still not sure. cred_show accepts a dict and build the string, that is ok because all values for credentials are strings so it is easy to generalize that. On the other hand, source_show has some interesting stuff in its output, it has integer and list also which makes a little harder to generalize.

You can ask, why don't you pass the dict as data and then use `json.dumps` to generate the output? That is a valid and a great question but the issue I found was to escape JSON structure for regex, like how I would differentiate that `[` is meant to match a literal `[` or be a regex list expression. Here is where things get interesting and I need to do a little more research if I will be able to do that.

With all those options in the table the last one way to generalize or avoid that repetition would be to have a format string constant for the output (as you suggested). Then call format and pass the specific data. This last option seems to be the one which won't mess with the regex escaping and JSON formatting.

As I said I will continue my research and see if I find another option that we should consider. Let me know if you have something else in your mind.

Reply:
That last option of having a format string constant seems the most sane!